### DISCUSSION TITLE
Fastify example?

### DISCUSSION THREAD

**Discussion by lockinmarv** [frandiox/vite-ssr#40] [Q&A] [answered]
Hey again!

Not sure if it's better to have a discussion rather an issue but I'm trying to get a server running with fastify.
Sadly it always fails when serving the static assets and I don't know why. 
The code is nearly the same as the express example:
```js
const fastify = require("fastify");
const fastifyStatic = require("fastify-static");
const path = require("path");

const { ssr } = require("./dist/server/package.json");
const manifest = require("./dist/client/ssr-manifest.json");
const { default: renderPage } = require("./dist/server");

const server = fastify({
  logger: false,
});

server.register(fastifyStatic, {
  root: __dirname,
});

for (const asset of ssr.assets || []) {
  server.get("/" + asset, (request, reply) => {
    reply.send("client/" + asset);
  });
}

server.get("*", async (request, reply) => {
  const url = request.protocol + "://" + request.headers.host + request.url;

  const { html } = await renderPage(url, {
    manifest,
    preload: true,
    request,
    response: reply,
  });

  reply.type("text/html").send(html);
});

startServer();

async function startServer() {
  try {
    await server.listen(3030);
  } catch (e) {
    console.error(e);
    process.exit(1);
  }
}
```
In the express example I need to serve the assets statically, that's why I thought I need to use `fastify-static` to do the same.
The difference here is that the `fastify-static` plugin needs a `root` where the `express.static` function doesn't. When running the server, it gives me an error because it just loads my `main.ts` (which is obviously not supported). It looks like it automatically serves all assets at the root, which includes the "raw" `index.html` too.

Specifying the root as `path.join(__dirname, 'dist')` doesn't work either. It prints `{"message":"Route GET:/ not found","error":"Not Found","statusCode":404}` in the browser.

Any idea what I need to change here to have the same behavior as the express server?

**Comment by frandiox** [comment, marked as answer]
@MarvinRudolph Hi! I think you just need to serve `dist/client` for static files. The way we do it in Express (1 by 1) is just to prevent serving `ssr-manifest.json` but I guess that's not a very big issue. Originally, it was also preventing serving the dummy `client/index.html` but this file is now removed so it shouldn't affect you.

Please tell me if that works 👍

**Reply by lockinmarv** [reply]
Ah, great! It works now after I've adjusted the `root` and added the `prefix` option. This code now works:
```ts
const fastify = require("fastify");
const fastifyStatic = require("fastify-static");
const path = require("path");
const manifest = require("./dist/client/ssr-manifest.json");
const { default: renderPage } = require("./dist/server");

const server = fastify({
  logger: false,
});

server.register(fastifyStatic, {
  root: path.join(__dirname, "dist", "client/assets"),
  prefix: "/assets",
});

server.get("*", async (request, reply) => {
  const url = request.protocol + "://" + request.headers.host + request.url;

  const { html } = await renderPage(url, {
    manifest,
    preload: true,
    request,
    response: reply,
  });

  reply.type("text/html").send(html);
});

startServer();

async function startServer() {
  try {
    await server.listen(3030);
  } catch (e) {
    console.error(e);
    process.exit(1);
  }
}
```
Thanks a lot! 🚀

**Reply by lockinmarv** [reply]
Any idea how I can serve files from the `public` directory? When I build, they get copied to `dist/client` which doesn't work with my current `dist/client/assets` configuration obviously. When serving from `dist/client` with the root set to `/`, the whole app can't be served.

**Reply by frandiox** [reply]
That's the normal behavior in Vite for the frontend I think. Can't you serve `path.join(__dirname, "dist", "client")` without any prefix (or prefix `/`)?
Otherwise maybe you can move files around after the build ends. Or put your files in `public/assets` so they will be copied to `client/assets` directly 🤔

**Reply by lockinmarv** [reply]
Neither serving from `dist/client` without the prefix nor with the prefix works. 😕 I also thought about moving files to `assets/` after the build, but wasn't sure if it's good to do that. Really frustrating with `fastify-static` where express just works. There must be something I've missed. 🤔😅

**Reply by frandiox** [reply]
@MarvinRudolph Did you figure it out? Fastify should support this in some way 🤔

**Reply by lockinmarv** [reply]
Sadly not yet. I gave up at some point but would be cool to find a solution for it. I could create a small repro where you can look into it if you're interested. I'm sure this should work but I just can't figure it out how. 😄

**Reply by frandiox** [reply]
I don't know much about Fastify really, perhaps they have a discord or Github Discussions?

**Reply by lockinmarv** [reply]
Yeah, I checked their Discord, but it isn't very active. Maybe I will try that again. :/
